### PR TITLE
Add Factorio

### PR DESCRIPTION
**Pull request by henriquegemignani**
Patcher: https://github.com/henriquegemignani/factorio-randovania-mod

Current Features:
- Full logic database, generated from data dumps exported from the game. Requirements have two parts:
   - Arbitrary "tier" for each research which leads to arbitrary requirements for each tier. Intended to provide some QoL research before you're required to automate high volume of things. 
   - Unlocked everything needed to research the necessary science packs
- Random recipe generation for science packs and rocket parts. These recipes are balanced to have a similar cost and complexity of the original recipes.
- Shuffles many of the start items and introduces a burner lab and assembling machine as starting items, to ensure it's always possible to automate science from start.
- Progressive pickups.

Todo (after this PR):
- Settings for tweaking random recipes (disable, only rocket parts, only science packs, maybe difficulty?)
- Better export dialog (find steam installation, find mod folder)
- Pickup pool with more copies of "bonus tech". Potentially balancing the tiers of these.
- Full playtesting and balancing

image
image
image

### CODE REVIEW

**Comment by Miepee** [comment]
tests fail

File: randovania/games/factorio/generator/pool_creator.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from randovania.games.factorio.layout.factorio_configuration import FactorioConfiguration
+
+if TYPE_CHECKING:
+    from randovania.game_description.game_description import GameDescription
+    from randovania.game_description.resources.resource_database import ResourceDatabase
+    from randovania.generator.pickup_pool import PoolResults
+    from randovania.layout.base.base_configuration import BaseConfiguration
+
+
+def create_victory_key(resource_database: ResourceDatabase) -> PickupEntry:
```
Comment (duncathan) [inline randovania/games/factorio/generator/pool_creator.py:17]: intended?

File: randovania/games/factorio/hash_words.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+HASH_WORDS = [
+    "Arugula",
```
Comment (duncathan) [inline randovania/games/factorio/hash_words.py:2]: proper hash words needed

File: randovania/lib/pyeasyga.py
Comment (duncathan) [inline randovania/lib/pyeasyga.py:1]: please add a comment indicating the changes made and why we can't use the pypi impl
(or make your own package tbh)

File: test/test_files/log_files/factorio/starter_preset.rdvgame [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+                                "num_shuffled_pickups": 1
+                            },
+                            "Physical Projectile Damage": {
+                                "num_shuffled_pickups": 1
+                            },
+                            "Refined Flammables": {
+                                "num_shuffled_pickups": 1
+                            },
+                            "Regular Inserter Capacity Bonus": {
+                                "num_shuffled_pickups": 1
```
Comment (duncathan) [inline test/test_files/log_files/factorio/starter_preset.rdvgame:92]: this appears to be an outdated starter preset. unsure if that's a problem

**Review by duncathan — CHANGES_REQUESTED** [review]

**Review by duncathan — APPROVED** [review]

**Pull request merged**